js/common/lib/training-session.ts

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import {InferenceSession} from './inference-session.js';
+import {TrainingSession as TrainingSessionImpl} from './training-session-impl.js';
+
+/* eslint-disable @typescript-eslint/no-redeclare */
+
+/**
+ * Represent a runtime instance of an ONNX model.
+ */
+export interface TrainingSession {
+  // #region run()
+
+  /**
+   * Run TrainStep asynchronously with the given feeds and options.
+   *
+   * @param feeds - Representation of the model input. See type description of `InferenceSession.InputType` for detail.
+   * @param options - Optional. A set of options that controls the behavior of model training.
+   * @returns A promise that resolves to an error code
+   */
+  runTrainStep(feeds: InferenceSession.FeedsType, options?: InferenceSession.RunOptions):
+      Promise<InferenceSession.ReturnType>;
+
+  /**
+   * Run a single train step with the given inputs and options.
+   *
+   * @param feeds - Representation of the model input.
+   * @param fetches - Representation of the model output.
+   * detail.
+   * @param options - Optional. A set of options that controls the behavior of model inference.
+   * @returns A promise that resolves to a map, which uses output names as keys and OnnxValue as corresponding values.
+   */
+  runTrainStep(
+      feeds: InferenceSession.FeedsType, fetches: InferenceSession.FetchesType,
+      options?: InferenceSession.RunOptions): Promise<InferenceSession.ReturnType>;
+
+  // #endregion
+
+  // #region release()
+
+  /**
+   * Release the inference session and the underlying resources.
+   */
+  release(): Promise<void>;
+
+  // #region metadata
+
+  /**
+   * Get input names of the loaded model.
+   */
+  readonly inputNames: readonly string[];
+
+  /**
+   * Get output names of the loaded model.
+   */
+  readonly outputNames: readonly string[];
+}
+
+export interface TrainingSessionFactory {
+  // #region create()
```

Comment:
The overrides are not defined in the best way.

so from number of parameter perspective, you can have the following possibliities:
1) checkpoint + trainModel
2) checkpoint + trainModel + sessionOptions
3) checkpoint + trainModel + evalModel
4) checkpoint + trainModel + evalModel + sessionOptions
5) checkpoint + trainModel + evalModel + optimizerModel
6) checkpoint + trainModel + evalModel + optimizerModel + sessionOptions


not sure if (3) and (4) are valid.

so you can use type `string|Uint8Array|ArrayBufferLike` as type of all parameters except sessionOptions. then you can only have 3 overrides ( or 2, if 3) and 4) above are invalid.)

Reply:
yes, I believe that all the possibilities listed are valid -- in federated learning scenarios, there's no need for the optimizerModel to be passed into the TrainingSession since those calculations/updates would be performed on a central server.

Reply:
These are the necessary overrides

1. checkpoint + trainModel
2. checkpoint + trainModel + sessionOptions
3. checkpoint + trainModel + optimizerModel
4. checkpoint + trainModel + optimizerModel+ sessionOptions
5. checkpoint + trainModel + evalModel + optimizerModel
6. checkpoint + trainModel + evalModel + optimizerModel + sessionOptions

In federated learning and sometimes in personalization too, eval step is not required. So we should have an option to build the training session without providing the eval model. 

Keeping optimizer model optional definitely gives users flexibility to only run train step but it is not required at this point. We can choose to keep the following 2 overloads or add them based on user demand:

checkpoint + trainModel + evalModel
checkpoint + trainModel + evalModel + sessionOptions

Reply:
I think it will be a little difficult to have all of the override possibilities represented through the parameters, since I believe that TypeScript will not be able to differentiate between the following:
* checkpoint + trainModel + optimizerModel
* checkpoint + trainModel + evalModel

The checkpoint + trainModel + optimizerModel possibility could be represented as: 
* checkpoint + trainModel + (empty Array or string) + optimizerModel 

& the documentation would specify that you can pass in an empty array or string, but I'm open to suggestions, since I think this solution is not so elegant. 

If the checkpoint + trainModel + optimizerModel combination is a lot more common than the checkpoint + trainModel + evalModel combination, we could also rearrange the parameters, but it comes with the trade-off of confusing users that are used to the other ORT-training APIs.

Reply:
For public interface, if you have multiple optional parameters, usually we use an option object to represent the parameters so that it is flexible for users.

you can define the type as this:
```ts
export interface TrainingSessionCreateOptions {
    checkpointState: string|Uint8Array|ArrayBufferLike;
    trainModel: string|Uint8Array|ArrayBufferLike;
    optimizerModel?: string|Uint8Array|ArrayBufferLike;
    evalModel?: string|Uint8Array|ArrayBufferLike;
    options?: InferenceSession.Options;
}
```

then the signature of function `create` can be simple:
```ts
  create(options: TrainingSessionCreateOptions): Promise<TrainingSession>;
```